### PR TITLE
Support open(WRONLY|RDWR) then truncate(0) in fuse

### PR DESCRIPTION
**Pull request by LuQQiu**
Modify for https://github.com/Alluxio/alluxio/issues/14895
### What changes are proposed in this pull request?
Fix open bugs
support open(WRONLY or RDWR) then truncate(0) then write() fuse case.
For Linux, it requires open(WRONLY or RDWR) on existing file with O_TRUNC to delete existing file.
But in libfuse, the workload can be open() then truncate(0).

OPEN(RDWR) will open for read-only first, if truncate(0) then write(), it will become write-only.

### Does this PR introduce any user facing changes?
Yeah, support open then truncate(0)

### CODE REVIEW

File: tests/src/test/java/alluxio/client/fuse/JNIFuseIntegrationTest.java [hunk 1]
```diff
@@ -43,4 +52,169 @@ public void mountFuse(FileSystem fileSystem, String mountPoint, String alluxioRo
...
+   */
+  @Test
+  public void openWrite() throws Exception {
+    String testFile = "/openWriteFile";
+    int fileLen = 128;
+    try (CloseableFuseFileInfo closeableFuseFileInfo = new CloseableFuseFileInfo()) {
+      FuseFileInfo info = closeableFuseFileInfo.getFuseFileInfo();
+      info.flags.set(OpenFlags.O_WRONLY.intValue());
+      // O_WRONLY non-existing file will write to file
+      Assert.assertEquals(0, mFuseFileSystem.open(testFile, info));
```
Comment (ssz1997) [inline tests/src/test/java/alluxio/client/fuse/JNIFuseIntegrationTest.java:86]: WorkerFuse doesn't create `mFuseFileSystem` upon mounting. There are test failures related to this

**Review by ssz1997 — CHANGES_REQUESTED** [review]
WorkerFuse won't pass the test. You can refer to github actions below for more info

File: tests/src/test/java/alluxio/client/fuse/JNIFuseIntegrationTest.java (hunk 1 above)
Comment (LuQQiu) [reply tests/src/test/java/alluxio/client/fuse/JNIFuseIntegrationTest.java:86]: Good catch, revert the inheritance, thanks

**Comment by LuQQiu** [comment]
https://github.com/Alluxio/alluxio/runs/5057544366?check_suite_focus=true
```
Error: 8.128 [ERROR] Failures: 
Error: 8.128 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.createWriteOpenRead
Error: 8.130 [ERROR]   Run 1: JNIFuseIntegrationTest.createWriteOpenRead:70->createTestFile:176 expected:<0> but was:<-5>
Error: 8.130 [ERROR]   Run 2: JNIFuseIntegrationTest.createWriteOpenRead:70->createTestFile:176 expected:<0> but was:<-5>
20:24:48.131 [INFO] 
Error: 8.131 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.openReadWrite
Error: 8.131 [ERROR]   Run 1: JNIFuseIntegrationTest.openReadWrite:130 expected:<0> but was:<-5>
Error: 8.132 [ERROR]   Run 2: JNIFuseIntegrationTest.openReadWrite:130 expected:<0> but was:<-5>
20:24:48.133 [INFO] 
Error: 8.133 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.openWrite
Error: 8.133 [ERROR]   Run 1: JNIFuseIntegrationTest.openWrite:86 expected:<0> but was:<-5>
Error: 8.133 [ERROR]   Run 2: JNIFuseIntegrationTest.openWrite:86 expected:<0> but was:<-5>
20:24:48.134 [INFO] 
```

```
Error: 8.060 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.openWrite  Time elapsed: 63.217 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<-5>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at alluxio.client.fuse.JNIFuseIntegrationTest.openWrite(JNIFuseIntegrationTest.java:86)
Error: 8.052 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.createWriteOpenRead  Time elapsed: 63.05 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<-5>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at alluxio.client.fuse.JNIFuseIntegrationTest.createTestFile(JNIFuseIntegrationTest.java:176)
	at alluxio.client.fuse.JNIFuseIntegrationTest.createWriteOpenRead(JNIFuseIntegrationTest.java:70)
```

**Comment by LuQQiu** [comment]
https://github.com/Alluxio/alluxio/runs/5057544313?check_suite_focus=true
```
Error: 1.355 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.openReadWrite  Time elapsed: 71.585 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<-5>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at alluxio.client.fuse.JNIFuseIntegrationTest.openReadWrite(JNIFuseIntegrationTest.java:130)
Error: 1.365 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.createWriteOpenRead  Time elapsed: 65.445 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<-5>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at alluxio.client.fuse.JNIFuseIntegrationTest.createTestFile(JNIFuseIntegrationTest.java:176)
Error: 1.372 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.openWrite  Time elapsed: 68.465 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<-5>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at alluxio.client.fuse.JNIFuseIntegrationTest.openWrite(JNIFuseIntegrationTest.java:86)
Error: 1.509 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.createWriteOpenRead
Error: 1.510 [ERROR]   Run 1: JNIFuseIntegrationTest.createWriteOpenRead:70->createTestFile:176 expected:<0> but was:<-5>
Error: 1.513 [ERROR]   Run 2: JNIFuseIntegrationTest.createWriteOpenRead:70->createTestFile:176 expected:<0> but was:<-5>
20:15:01.515 [INFO] 
Error: 1.515 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.openReadWrite
Error: 1.516 [ERROR]   Run 1: JNIFuseIntegrationTest.openReadWrite:130 expected:<0> but was:<-5>
Error: 1.516 [ERROR]   Run 2: JNIFuseIntegrationTest.openReadWrite:130 expected:<0> but was:<-5>
20:15:01.516 [INFO] 
Error: 1.517 [ERROR] alluxio.client.fuse.JNIFuseIntegrationTest.openWrite
Error: 1.517 [ERROR]   Run 1: JNIFuseIntegrationTest.openWrite:86 expected:<0> but was:<-5>
Error: 1.517 [ERROR]   Run 2: JNIFuseIntegrationTest.openWrite:86 expected:<0> but was:<-5>
20:15:01.517 [INFO] 
```

**Comment by LuQQiu** [comment]
https://github.com/Alluxio/alluxio/runs/5057544360?check_suite_focus=true
```
Error: 9.878 [ERROR] Errors: 
Error: 9.879 [ERROR] alluxio.job.master.JobMasterIntegrationTest.cancel
Error: 9.879 [ERROR]   Run 1: JobMasterIntegrationTest.cancel:237 expected:<3> but was:<1>
Error: 9.879 [ERROR]   Run 2: JobMasterIntegrationTest.cancel:244 » Timeout Timed out waiting for job 164391...
20:23:49.879 [INFO] 
Error: 9.879 [ERROR] alluxio.job.master.JobMasterIntegrationTest.multipleTasksPerWorker
Error: 9.879 [ERROR]   Run 1: JobMasterIntegrationTest.multipleTasksPerWorker:91 » Timeout Timed out waiting...
Error: 9.879 [ERROR]   Run 2: JobMasterIntegrationTest.multipleTasksPerWorker:91 » Timeout Timed out waiting...
20:23:49.880 [INFO] 
Error: 9.880 [ERROR] alluxio.job.master.JobMasterIntegrationTest.restartMasterAndReregisterWorker
Error: 9.880 [ERROR]   Run 1: JobMasterIntegrationTest.restartMasterAndReregisterWorker:133 » Timeout Timed ...
Error: 9.880 [ERROR]   Run 2: JobMasterIntegrationTest.restartMasterAndReregisterWorker:133 » Timeout Timed ...
20:23:49.880 [INFO] 
Error: 9.880 [ERROR] alluxio.job.master.JobMasterIntegrationTest.stopJobWorkerTasks
Error: 9.880 [ERROR]   Run 1: JobMasterIntegrationTest.stopJobWorkerTasks:172 » Timeout Timed out waiting fo...
Error: 9.880 [ERROR]   Run 2: JobMasterIntegrationTest.stopJobWorkerTasks:172 » Timeout Timed out waiting fo...
20:23:49.881 [INFO] 
Error: 9.881 [ERROR] alluxio.job.master.JobMasterIntegrationTest.throttleJobWorkerTasks
Error: 9.881 [ERROR]   Run 1: JobMasterIntegrationTest.throttleJobWorkerTasks:197 » Timeout Timed out waitin...
Error: 9.881 [ERROR]   Run 2: JobMasterIntegrationTest.throttleJobWorkerTasks:197 » Timeout Timed out waitin...
```

**Comment by LuQQiu** [comment]
@ssz1997 Remove the tests from this PR and focus on the logic fix first.
The test cases include in previous commits include the open then truncate case. That case is validated locally.

**Comment by LuQQiu** [comment]
@ssz1997 PTAL, thanks

**Review by ssz1997 — APPROVED** [review]

**Comment by LuQQiu** [comment]
https://github.com/Alluxio/alluxio/runs/5069014896?check_suite_focus=true
```
Error: 0.799 [ERROR] Failures: 
Error: 0.800 [ERROR] alluxio.master.AlluxioMasterProcessTest.stopAfterStandbyTransition[1]
Error: 0.800 [ERROR]   Run 1: AlluxioMasterProcessTest.stopAfterStandbyTransition:209
Error: 0.801 [ERROR]   Run 2: AlluxioMasterProcessTest.stopAfterStandbyTransition:209
16:18:10.805 [INFO] 
```

**Comment by LuQQiu** [comment]
Error: 3.469 [ERROR] Tests run: 2, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 70.467 s <<< FAILURE! - in alluxio.client.fs.concurrent.ConcurrentFileInStreamIntegrationTest
Error: 3.476 [ERROR] alluxio.client.fs.concurrent.ConcurrentFileInStreamIntegrationTest.FileInStreamConcurrency  Time elapsed: 30.688 s  <<< FAILURE!
java.lang.AssertionError: Failed with exception(s) [java.lang.RuntimeException: java.io.IOException: Failed to connect to remote block worker: 
https://github.com/Alluxio/alluxio/runs/5069015154?check_suite_focus=true

**Comment by LuQQiu** [comment]
alluxio-bot, merge this please

**Pull request merged**